src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```diff
@@ -69,16 +89,24 @@ public int Take()
             }
 
             var index = _index;
-            var wasLastBlock = block.Next == null;
 
             if (index < block.End)
             {
                 _index = index + 1;
                 return block.Array[index];
             }
 
+            return TakeMultiBlock();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private int TakeMultiBlock()
+        {
+            var block = _block;
+            var wasLastBlock = block.Next == null;
```

Comment:
Yeah. I probably should have done that in the first place. Hopefully we can come up with a shorter version of that comment tough.